org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1279 @@
+            // be created by VersionRangeGet(...).
+            throw new RuntimeException("JSSEngine.queryEnabledProtocols() - null protocol range");
+        }
+
+        min_protocol = vrange.getMinVersion();
+        max_protocol = vrange.getMaxVersion();
+    }
+
+    public String[] getEnabledProtocols() {
+        debug("JSSEngine: getEnabledProtocols()");
+
+        if ((min_protocol == null || max_protocol == null) && ssl_fd != null) {
+            queryEnabledProtocols();
+        }
+
+        if (min_protocol == null || max_protocol == null) {
+            // TODO: Query default ciphersuites here.
+            throw new RuntimeException("JSSEngine.getEnabledProtocls() - Unable to query enabled protocols off of empty ssl_fd.");
+        }
+
+        // Use JSSParameters to do the heavy lifting of converting the
+        // SSLVersionRange to a list of JDK-conforming Strings.
+        JSSParameters parser = new JSSParameters();
+        parser.setProtocols(min_protocol, max_protocol);
+        return parser.getProtocols();
+    }
+
+    public boolean getEnableSessionCreation() {
+        debug("JSSEngine: getEnableSessionCreation() - not implemented");
+        return true;
+    }
+
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        debug("JSSEngine: getHandshakeStatus()");
+        try {
+            updateHandshakeState();
+        } catch (SSLException se) {
+            // Do nothing -- updateHandshakeStatus() should throw the same
+            // exception again, so it is safe to discard it. We also can't
+            // just pass the SSLException to the caller of this method, as
+            // they aren't expecting it.
+        }
+
+        return handshake_state;
+    }
+
+    public SSLSession getSession() {
+        debug("JSSEngine: getSession()");
+        return session;
+    }
+
+    public String[] getSupportedCipherSuites() {
+        debug("JSSEngine: getSupportedCipherSuites()");
+        ArrayList<String> result = new ArrayList<String>();
+
+        for (SSLCipher c : SSLCipher.values()) {
+            if (c.isSupported()) {
+                debug("JSSEngine: getSupportedCipherSuites() - Supported: " + c);
+                result.add(c.name());
+            }
+        }
+
+        return result.toArray(new String[result.size()]);
+    }
+
+    public String[] getSupportedProtocols() {
+        debug("JSSEngine: getSupportedProtocols()");
+        ArrayList<String> result = new ArrayList<String>();
+
+        for (SSLVersion v : Policy.TLS_VERSION_RANGE.getAllInRange()) {
+            debug("JSSEngine: getSupportedProtocol - Supported: " + v);
+            result.add(v.jdkAlias());
+        }
+
+        return result.toArray(new String[result.size()]);
+    }
+
+    public SecurityStatusResult getStatus() {
+        if (ssl_fd == null) {
+            return null;
+        }
+
+        return SSL.SecurityStatus(ssl_fd);
+    }
+
+    public void setEnableSessionCreation(boolean flag) {
```

Comment:
Should move this to JSSEngine. All JSSEngines will have the same behavior here likely. I don't think we can restrict to only resuming existing sessions.